proposals/NNNN-DXIL-Scalarization.md

```diff
@@ -0,0 +1,90 @@
+# DXIL Scalarization
+
+* Proposal: [NNNN](NNNN-DXIL-Scalarization.md)
+* Author(s): [Farzon Lotfi](https://github.com/farzonl)
+* Sponsor: [Farzon Lotfi](https://github.com/farzonl)
+* Status: **Under Consideration**
+* Impacted Projects: Clang
+
+## Introduction
+As mentioned in [DXIL.rst](https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/DXIL.rst#vectors) 
+"HLSL vectors are scalarized" and "Matrices are lowered to vectors". Therefore,
+ we need to be able to scalarize datastructures, call instructions, and vector 
+ operations like mathamatical, logical and bitcasts. The goal of this proposal 
+ is to present a solution for LLVM's pass manager.
+
+## Motivation
+Without Scalarizing the data structures and call instructions we can't generate
+ valid DXIL.
+
+ ## Background
+ In DXC Scalarization and SROA are merged in unfortnate ways. Most of that code
+ can be found here: [`SROA_Parameter_HLSL` Module pass in `ScalarReplAggregatesHLSL.cpp`](https://github.com/microsoft/DirectXShaderCompiler/blob/main/lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp#L4263). 
+The specific instructions DXC operates on are `CallInst`, `LoadInst`, 
+`StoreInst`, `GetElementPtrInst`, `AddrSpaceCastInst`, `BitCastInst`, and 
+`MemIntrinsic`. Not covered in this proposal is the handling of `Allocas`.
+That will be addressed in a later proposal covering DXIL Legalization.
+
+## Proposal
+Our goal should be legal DIXL with behavioral similarity to DXC. To that end it
+ makes the most sense to use the Scalarizer pass.
+- [Scalarizer.cpp](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Scalar/Scalarizer.cpp)
+
+The `scalarizer` pass with the `-scalarize-load-store` and 
+`-scalarize-variable-insert-extract` flags  will handle the `BitCastInst`, 
+`CallInst`, `LoadInst` and `StoreInst`, and `GetElementPtrInst` cases. The 
+`AddrSpaceCastInst` case isn't relevant for upstream LLVM.
+
+`MemIntrinsic` is the odd case out. We don't currently emit memcpy for
+the same cases we do in DXC.
+In many cases operations like memcpy and memset seem to happen in global scope.
```

Comment:
> In many cases operations like memcpy and memset seem to happen in global scope.

In DXC or in clang? Or both? For the godbolt link below I only see memcpy in the dxc -fcgl version.

Reply:
In DXC.  I'll specify DXC in the paragraph. I just didn't generate the other cases, but the code that inserts `memcpy` aslo handles the `memset` and `memmove`  emits. It looks like this:
```cpp
// Finally, insert the meminst for this element.
    if (isa<MemSetInst>(MI)) {
      Builder.CreateMemSet(EltPtr, MI->getArgOperand(1), EltSize,
                           MI->isVolatile());
    } else {
      assert(isa<MemTransferInst>(MI));
      Value *Dst = SROADest ? EltPtr : OtherElt; // Dest ptr
      Value *Src = SROADest ? OtherElt : EltPtr; // Src ptr

      if (isa<MemCpyInst>(MI))
        Builder.CreateMemCpy(Dst, Src, EltSize, OtherEltAlign,
                             MI->isVolatile());
      else
        Builder.CreateMemMove(Dst, Src, EltSize, OtherEltAlign,
                              MI->isVolatile());
    }
```
So DXC inserts the Mem intrinsics as part of `RewriteMemIntrin`